Speaker Feedback: Add UI for speakers to opt out of email notifications about approved feedback

coreymckrill opened the pull request:
Building on the usermeta value introduced in #441, this adds a toggle to the frontend Session Feedback view where speakers and organizers can disable and re-enable email notifications about newly approved feedback submissions. The toggle updates the usermeta value, which carries over across all WordCamp sites in the network.

To accomplish this, I created a new endpoint on our custom REST API route for feedback. The standard `users` endpoint doesn't work for updating this usermeta value because it requires that the user be a member on the current site. Most speakers are not added as members on the WordCamp sites that they speak at. The custom endpoint allows us to circumvent that.

Note that this does **not** include a metabox or other UI on the Speaker CPT screen for toggling this value. Adding a metabox that would be functional in the block editor turned out to be overly-complex, given the necessary interaction with the existing `WordPress.org Username` metabox field. Building a custom sidebar plugin is beyond the scope of this PR, but it might be simpler, and doable as a separate PR, once the other metabox fields for the Speaker CPT have been converted to sidebar panels, as was [recently done](https://github.com/WordPress/wordcamp.org/pull/303) for the Sessions CPT. In the mean time, organizers can toggle this value for speakers upon request by going to the frontend feedback view, where they will see the same UI that speakers do, but with a separate toggle for each speaker in the session.

Fixes #488 

### Screenshots

Speaker UI, notifications enabled (default)

![speaker-view-enabled](https://user-images.githubusercontent.com/916023/81439825-d912eb00-9123-11ea-935e-5255138ca204.jpg)

Speaker UI, notifications disabled

![speaker-view-disabled](https://user-images.githubusercontent.com/916023/81439837-e203bc80-9123-11ea-86ca-ba8c111ac7d6.jpg)

Organizer UI

![admin-view](https://user-images.githubusercontent.com/916023/81453076-7c242e80-913d-11ea-8351-338b46f337b3.jpg)

### How to test the changes in this Pull Request:

1. Rebuild the stylesheet to get the style updates.
1. With the same testing setup as for #441, open the feedback moderation screen and approve a new feedback submission.
1. Run the cron job and note that an email is sent in MailCatcher.
1. In a separate tab, navigate to the frontend feedback view for a session while logged in as a speaker. Click the toggle to disable notifications.
1. Approve another feedback submission and run the cron job. No new email should be sent.
1. Re-enable notifications for the speaker, and run the cron job again. Now the email should get sent.

coreymckrill committed:
Add a REST endpoint for updating notifications

coreymckrill committed:
Add UI to feedback view for toggling notifications

coreymckrill committed:
Fix css linting issue

coreymckrill committed:
Change notification toggle UI to allow use by admins

coreymckrill committed:
Remove code comment

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/views/view-feedback.php:
```diff
@@ -41,6 +44,51 @@
+		<div class="speaker-feedback__notifications <?php echo ( $notifications_opt_out ) ? 'is-disabled' : ''; ?>">
+			<span>
+				<?php if ( ! $is_session_speaker ) : ?>
+					<span class="sft-notifications-speaker-name">
```
I know it's long, but these should follow the class name patterns: `speaker-feedback__notifications-speaker-name`, `speaker-feedback__notifications-label-text`, etc. (`sft-` is only used on IDs/names)

coreymckrill replied:
Fixed with fe3c145

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/views/view-feedback.php:
```diff
@@ -41,6 +44,51 @@
+				<input
+					type="checkbox"
+					data-user-id="<?php echo absint( $user->ID ); ?>"
+					<?php checked( $notifications_opt_out ); ?>
```
This needs an `aria-describedby` pointing to the content in the visual label, otherwise it's not totally clear to a screen reader user what you're enabling.

coreymckrill replied:
Fixed with 4fe2df0

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/class-rest-feedback-controller.php:
```diff
@@ -67,6 +68,36 @@ public function register_routes() {
+		// so we have to create our own endpoint.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/notifications/(?P<id>[\d]+)',
```
I think this endpoint needs a separate controller, `class-rest-notifications-controller.php`, so that it can be its own base: `/wordcamp-speaker-feedback/v1/notifications/…`

It would also be nice to see tests for this endpoint (that doesn't have to block merge though)

coreymckrill replied:
Sigh. You're right.

Moved to separate controller here: 40b5d8f

Tests added here: e836d56

ryelle reviewed (commented):
It works as described, toggling off notifs skipped that user in the automated emails. It's a little unclear with a screen reader, but that should be a simple fix (noted below).

I feel like this needs some help text, maybe above the box. Thinking for speakers, something like:
> You will receive up to an email a day if there are new feedback submissions. Disabling this will turn off notifications for all feedback.

And organizers:
> Manage speaker notifications. Speakers will receive up to an email a day if there are new feedback submissions.

coreymckrill committed:
Tweak wording of notices

coreymckrill committed:
Fix class names to match project patterns

coreymckrill committed:
Add aria-describedby attribute

coreymckrill committed:
Move notifications endpoint to separate REST controller

coreymckrill committed:
Enable single site admins to pass the permissions check

On multisite, single site admins don't have the `edit_user` cap.
We want them to be able to toggle notifications for speakers, though.
As an alternative, we can use the existing `moderate_COMMENT_TYPE` cap
instead.

coreymckrill committed:
Add and update tests

coreymckrill commented:
Helpful feedback on this, thanks. I've added help text blurbs. Didn't go with exactly what you had in mind, the wording is here: b150ad7

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/views/view-feedback.php:
```diff
@@ -41,6 +44,69 @@
+		$notifications_opt_out = get_user_meta( $user->ID, SPEAKER_OPT_OUT_KEY, true );
+		?>
+		<div class="speaker-feedback__notifications <?php echo ( $notifications_opt_out ) ? 'is-disabled' : ''; ?>">
+			<span id="sft-notifications-description">
```
This isn't unique in the organizer's view if there's more than one speaker. appending the speaker ID should be good to make it unique.

coreymckrill replied:
Thanks, good catch. 0b4d5a90

ryelle reviewed (approved):
One minor note that should get fixed before merging, but otherwise it's working great 👍

coreymckrill committed:
Make the notifs description id unique to accommodate more than one on a page
